lib/fluent/plugin/in_tail.rb

```diff
@@ -498,36 +495,37 @@ def on_rotate(io)
               pos = @read_from_head ? 0 : fsize
               @pe.update(inode, pos)
             end
-            io.seek(pos)
-
-            @io_handler = IOHandler.new(io, @pe, @log, @read_lines_limit, &method(:wrap_receive_lines))
+            @io_handler = IOHandler.new(self, &method(:wrap_receive_lines))
           else
             @io_handler = NullIOHandler.new
           end
         else
-          log_msg = "detected rotation of #{@path}"
-          log_msg << "; waiting #{@rotate_wait} seconds" if @io_handler.io  # wait rotate_time if previous file is exist
-          @log.info log_msg
+          watcher_needs_update = false
 
-          if io
-            stat = io.stat
+          if stat
             inode = stat.ino
             if inode == @pe.read_inode # truncated
-              @pe.update_pos(stat.size)
-              io_handler = IOHandler.new(io, @pe, @log, @read_lines_limit, &method(:wrap_receive_lines))
+              @pe.update_pos(0)
```

Comment:
The current behavior is somewhat strange.  Because the file position of `io`, which must have been newly opened by the caller, isn't updated while the position is updated to `stat.size` here, IOHandler later starts reading from the head anyway.

The new code tries to update the file handle's position to that stored in `@pe` before reading from the target file, so the position has to be set to 0 in order to preserve the behavior.

If this is a bug, it needs to be corrected to `@pe.update_pos(stat.size)`.